diff -rua qt-x11-free-3.3.3.old/src/kernel/qscriptengine_x11.cpp qt-x11-free-3.3.3/src/kernel/qscriptengine_x11.cpp
--- qt-x11-free-3.3.3.old/src/kernel/qscriptengine_x11.cpp	2004-08-05 20:42:08.000000000 +0600
+++ qt-x11-free-3.3.3/src/kernel/qscriptengine_x11.cpp	2004-09-26 12:30:55.000000000 +0600
@@ -522,7 +522,7 @@
     Invalid, Consonant, Invalid, Invalid,
 
     Consonant, Consonant, Consonant, Consonant,
-    Consonant, Consonant, Invalid, Invalid,
+    Consonant, Consonant, Consonant, Invalid,
     Invalid, Invalid, Halant, Invalid,
     Invalid, Invalid, Invalid, Matra,
 
@@ -1092,6 +1092,14 @@
     { (Form)0, None }
 };
 
+static const IndicOrdering sinhala_order [] = {
+    { Matra, Below },
+    { Matra, Above },
+    { Matra, Post },
+    { VowelMark, Post },
+    { (Form)0, None }
+};
+
 static const IndicOrdering * const indic_order[] = {
     devanagari_order, // Devanagari
     bengali_order, // Bengali
@@ -1102,7 +1110,7 @@
     telugu_order, // Telugu
     kannada_order, // Kannada
     malayalam_order, // Malayalam
-    devanagari_order // Sinhala // ### no OT specs available, we use devanagari
+    sinhala_order // Sinhala
 };
 
 
@@ -1725,7 +1733,8 @@
 		break;
 	    goto finish;
 	case Consonant:
-	    if ( state == Halant )
+	    if ( state == Halant && (( script != QFont::Sinhala )
+		|| ( uc[pos-1].unicode() == 0x200d )))
 		break;
 	    goto finish;
 	case Halant:
