diff -Nrua pango-1.5.2/modules/indic/indic-fc.c pango-1.5.2.new/modules/indic/indic-fc.c
--- pango-1.5.2/modules/indic/indic-fc.c	2004-08-02 16:58:31.000000000 -0400
+++ pango-1.5.2.new/modules/indic/indic-fc.c	2004-08-11 13:07:30.000000000 -0400
@@ -65,6 +65,7 @@
 #define OT_TAG_telu FT_MAKE_TAG('t','e','l','u')
 #define OT_TAG_knda FT_MAKE_TAG('k','n','d','a')
 #define OT_TAG_mlym FT_MAKE_TAG('m','l','y','m')
+#define OT_TAG_sinh FT_MAKE_TAG('s','i','n','h')
 
 static PangoEngineScriptInfo deva_scripts[] = {
   { PANGO_SCRIPT_DEVANAGARI, "*" }
@@ -102,10 +103,15 @@
   { PANGO_SCRIPT_MALAYALAM, "*" }
 };
 
+static PangoEngineScriptInfo sinh_scripts[] = {
+  { PANGO_SCRIPT_SINHALA, "*" }
+};
+
 static PangoEngineInfo script_engines[] = {
   INDIC_ENGINE_INFO(deva), INDIC_ENGINE_INFO(beng), INDIC_ENGINE_INFO(guru),
   INDIC_ENGINE_INFO(gujr), INDIC_ENGINE_INFO(orya), INDIC_ENGINE_INFO(taml),
-  INDIC_ENGINE_INFO(telu), INDIC_ENGINE_INFO(knda), INDIC_ENGINE_INFO(mlym)
+  INDIC_ENGINE_INFO(telu), INDIC_ENGINE_INFO(knda), INDIC_ENGINE_INFO(mlym),
+  INDIC_ENGINE_INFO(sinh)
 };
 
 /*
@@ -119,7 +125,8 @@
 static PangoIndicInfo indic_info[] = {
   PANGO_INDIC_INFO(deva), PANGO_INDIC_INFO(beng), PANGO_INDIC_INFO(guru),
   PANGO_INDIC_INFO(gujr), PANGO_INDIC_INFO(orya), PANGO_INDIC_INFO(taml),
-  PANGO_INDIC_INFO(telu), PANGO_INDIC_INFO(knda), PANGO_INDIC_INFO(mlym)
+  PANGO_INDIC_INFO(telu), PANGO_INDIC_INFO(knda), PANGO_INDIC_INFO(mlym),
+  PANGO_INDIC_INFO(sinh)
 };
 
 static void
diff -Nrua pango-1.5.2/modules/indic/indic-ot-class-tables.c pango-1.5.2.new/modules/indic/indic-ot-class-tables.c
--- pango-1.5.2/modules/indic/indic-ot-class-tables.c	2004-08-02 14:43:09.000000000 -0400
+++ pango-1.5.2.new/modules/indic/indic-ot-class-tables.c	2004-08-11 13:07:30.000000000 -0400
@@ -192,6 +192,19 @@
     _iv, _iv, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx  /* 0D60 - 0D6F */
 };
 
+static IndicOTCharClass sinhCharClasses[] =
+{
+    _xx, _xx, _mp, _mp, _xx, _iv, _iv, _iv, _iv, _iv, _iv, _iv, _iv, _iv, _iv, _iv, /* 0D80 - 0D8F */
+    _iv, _iv, _iv, _iv, _iv, _iv, _iv, _xx, _xx, _xx, _ct, _ct, _ct, _ct, _ct, _ct, /* 0D90 - 0D9F */
+    _ct, _ct, _ct, _ct, _ct, _ct, _ct, _ct, _ct, _ct, _ct, _ct, _ct, _ct, _ct, _ct, /* 0DA0 - 0DAF */
+    _ct, _ct, _xx, _ct, _ct, _ct, _ct, _ct, _ct, _ct, _ct, _ct, _xx, _ct, _xx, _xx, /* 0DB0 - 0DBF */
+    _ct, _ct, _ct, _ct, _ct, _ct, _ct, _xx, _xx, _xx, _vr, _xx, _xx, _xx, _xx, _dr, /* 0DC0 - 0DCF */
+    _dr, _dr, _da, _da, _db, _xx, _db, _xx, _dr, _dl, _s1, _dl, _s2, _s3, _s4, _dr, /* 0DD0 - 0DDF */
+    _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, _xx, /* 0DE0 - 0DEF */
+    _xx, _xx, _dr, _dr, _xx                                                         /* 0DF0 - 0DF4 */
+};
+
+
 /*
  * Split matra tables
  */
@@ -208,6 +221,9 @@
 
 static const IndicOTSplitMatra mlymSplitTable[] = {{0x0D46, 0x0D3E}, {0x0D47, 0x0D3E}, {0x0D46, 0x0D57}};
 
+static const IndicOTSplitMatra sinhSplitTable[] = {{0x0DD9, 0x0DCA}, {0x0DD9, 0x0DCF}, {0x0DD9, 0x0DCF,0x0DCA}, {0x0DD9, 0x0DDF} };
+
+
 /*
  * Script Flags
  */
@@ -226,6 +242,7 @@
 #define TELU_SCRIPT_FLAGS (SF_MATRAS_AFTER_BASE | 3)
 #define KNDA_SCRIPT_FLAGS (SF_MATRAS_AFTER_BASE | 3)
 #define MLYM_SCRIPT_FLAGS (SF_MPRE_FIXUP | SF_NO_POST_BASE_LIMIT)
+#define SINH_SCRIPT_FLAGS (SF_MPRE_FIXUP | SF_NO_POST_BASE_LIMIT)
 
 /*
  * Indic Class Tables
@@ -248,6 +265,8 @@
 
 IndicOTClassTable mlym_class_table = {0x0D00, 0x0D6F, 3, MLYM_SCRIPT_FLAGS, mlymCharClasses, mlymSplitTable};
 
+IndicOTClassTable sinh_class_table = {0x0D80, 0x0DF4, 3, SINH_SCRIPT_FLAGS, sinhCharClasses, sinhSplitTable};
+
 const IndicOTSplitMatra *indic_ot_get_split_matra(const IndicOTClassTable *class_table, IndicOTCharClass char_class)
 {
     gint32 index = (char_class & CF_INDEX_MASK) >> CF_INDEX_SHIFT;
diff -Nrua pango-1.5.2/modules/indic/indic-ot.c pango-1.5.2.new/modules/indic/indic-ot.c
--- pango-1.5.2/modules/indic/indic-ot.c	2004-08-02 16:58:31.000000000 -0400
+++ pango-1.5.2.new/modules/indic/indic-ot.c	2004-08-11 13:08:35.000000000 -0400
@@ -29,6 +29,7 @@
     gunichar fMabove;
     gunichar fMpost;
     gunichar fLengthMark;
+    gunichar fVirama; // to handle virama in sinhala split matras
     glong    fMatraIndex;
     gulong   fMatraTags;
     gboolean fMatraWordStart;
@@ -50,7 +51,7 @@
     output->fOutIndex    = 0;
     output->fMatraTags   = 0;
 
-    output->fMpre = output->fMbelow = output->fMabove = output->fMpost = output->fLengthMark = 0;
+    output->fMpre = output->fMbelow = output->fMabove = output->fMpost = output->fLengthMark = output->fVirama = 0;
 
     output->fMPreOutIndex = -1;
     output->fMPreFixups = mpreFixups;
@@ -69,12 +70,14 @@
         output->fMpost = matra;
     } else if (IS_LENGTH_MARK(matraClass)) {
         output->fLengthMark = matra;
+    } else if(IS_VIRAMA(matraClass)) {
+        output->fVirama = matra;
     }
 }
 
 static void initMatra(Output *output, guint32 matraIndex, gulong matraTags, gboolean wordStart)
 {
-    output->fMpre = output->fMbelow = output->fMabove = output->fMpost = output->fLengthMark = 0;
+    output->fMpre = output->fMbelow = output->fMabove = output->fMpost = output->fLengthMark = output->fVirama = 0;
     output->fMPreOutIndex = -1;
     output->fMatraIndex = matraIndex;
     output->fMatraTags = matraTags;
@@ -122,6 +125,14 @@
     output->fOutIndex += 1;
 }
 
+//To handle virama in sinhala split matra
+static void writeVirama(Output *output)
+{
+  if(output->fVirama != 0){
+    writeChar(output, output->fVirama, output->fMatraIndex, output->fMatraTags);
+  }
+}
+
 static void writeMpre(Output *output)
 {
     if (output->fMpre != 0) {
@@ -230,6 +241,7 @@
             writeMabove(&output);
             writeMpost(&output);
             writeLengthMark(&output);
+	    writeVirama(&output);
             break;
 
         case CC_CONSONANT:
@@ -414,6 +426,7 @@
 	    }
 
 	    writeLengthMark(&output);
+	    writeVirama(&output);
 
 	    /* write reph */
 	    if ((class_table->scriptFlags & SF_REPH_AFTER_BELOW) == 0) {
diff -Nrua pango-1.5.2/modules/indic/indic-ot.h pango-1.5.2.new/modules/indic/indic-ot.h
--- pango-1.5.2/modules/indic/indic-ot.h	2004-08-02 16:58:31.000000000 -0400
+++ pango-1.5.2.new/modules/indic/indic-ot.h	2004-08-11 13:07:30.000000000 -0400
@@ -197,6 +197,7 @@
 extern IndicOTClassTable telu_class_table;
 extern IndicOTClassTable knda_class_table;
 extern IndicOTClassTable mlym_class_table;
+extern IndicOTClassTable sinh_class_table;
 
 const IndicOTSplitMatra *indic_ot_get_split_matra(const IndicOTClassTable *class_table, IndicOTCharClass char_class);
 
